### PR TITLE
[NADE] Add processor entry to Artifact Mapping in PDF

### PR DESCRIPTION
**Pull request by sfc-gh-bgoel**
### Pre-review checklist
   * [x] I've confirmed that instructions included in README.md are still correct after my changes in the codebase.
   * [x] I've added or updated automated unit tests to verify correctness of my new code.
   * [ ] I've added or updated integration tests to verify correctness of my new code.
   * [ ] I've confirmed that my changes are working by executing CLI's commands manually.
   * [x] I've confirmed that my changes are up-to-date with the target branch - as of April 29, 8AM EST
   * [ ] I've described my changes in the release notes.
   * [x] I've described my changes in the section below.

### Changes description
As part of setup script generation, we need to allow specifying processors for artifacts which will discover and work on annotated code. This PR is one of many PRs.

### CODE REVIEW

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 1]
```diff
@@ -1,10 +1,43 @@
 from __future__ import annotations
 
-from typing import Optional
+from typing import List, Literal, Optional
 
+from pydantic import Field
 from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel
 
+# Limited to python options only for now.
+ProcessorOptions = Literal["python-snowpark", "PYTHON-SNOWPARK"]
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:9]: I'd use snowpark-python here, or even just snowpark. We can infer it's python (vs java later on).

Also, the plugin infrastructure will require this to be a string. I'd bake that in upfront and just have dynamic validation of the processor name with the one supported value later in the pipeline.

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 2]
```diff
@@ -1,10 +1,43 @@
...
+    env_type: VirtualEnvTypeOptions = Field(
+        title="Type of the virtual environment to activate.",
+        default="venv",
+    )
+
+
+class AnnotationProcessor(UpdatableModel):
+    language: LanguageOptions = Field(
+        title="Programming language in which the annotated code is written.",
+    )
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:27]: I don't think we need this

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 3]
```diff
@@ -1,10 +1,43 @@
...
+    language_version: str = Field(
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:28]: Do we need this? IIRC snowpark uses the annotations for this, or defaults to the version of python that we'll be invoking to run the user's code.

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 4]
```diff
@@ -1,10 +1,43 @@
...
+        title="Programming language version in which the annotated code is written, and to use in a virtual environment, if provided.",
+    )
+    name: ProcessorOptions = Field(
+        title="Name of the processor to be invoked to discover the annotated code."
+    )
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:33]: We should support having just a string as processor (interpreted as the name)

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 5]
```diff
@@ -1,10 +1,43 @@
...
 
-from typing import Optional
+from typing import List, Literal, Optional
 
+from pydantic import Field
 from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel
 
+# Limited to python options only for now.
+ProcessorOptions = Literal["python-snowpark", "PYTHON-SNOWPARK"]
+VirtualEnvTypeOptions = Literal["conda", "CONDA", "venv", "VENV"]
```
Comment (sfc-gh-cgorrie) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:10]: Nit: do we need to support the uppercase versions of these?

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 6]
```diff
@@ -1,10 +1,36 @@
...
+VirtualEnvTypeOptions = Literal["conda", "venv"]
+
+
+class VirtualEnvironment(UpdatableModel):
+    name: str = Field(
+        title="Name of the virtual environment to run the processor in.",
+    )
+    env_type: VirtualEnvTypeOptions = Field(
+        title="Type of the virtual environment to activate.",
+        default="venv",
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:19]: there is no reasonable default, we should require this field

File: tests/test_data/projects/napp_with_annotation_processor/snowflake.yml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+  artifacts:
+    - setup.sql
+    - app/README.md
+    - src: app/streamlit/*.py
+      dest: ui/
+      processors:
+        - snowpark # str instead of dict
+        - name: snowpark # without virtual_env
+        - name: snowpark # with virtual_env
```
Comment (sfc-gh-bdufour) [inline tests/test_data/projects/napp_with_annotation_processor/snowflake.yml:17]: maybe vary them a little? Eventually we'll likely consider this an invalid PDF due to repetition

**Review by sfc-gh-astus — DISMISSED** [review]

**Review by sfc-gh-astus — APPROVED** [review]

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 8]
```diff
@@ -1,10 +1,22 @@
...
-from typing import Optional
+from typing import Any, Dict, List, Optional, Union
 
+from pydantic import Field
 from snowflake.cli.api.project.schemas.updatable_model import UpdatableModel
 
 
+class AnnotationProcessor(UpdatableModel):
+    name: str = Field(
+        title="Name of the processor to be invoked to discover the annotated code."
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:11]: nit: "Name of a processor to invoke on a collection of artifacts" ? Annotation processing is a special case.

File: src/snowflake/cli/api/project/schemas/native_app/path_mapping.py [hunk 9]
```diff
@@ -1,10 +1,22 @@
...
+    )
+    properties: Optional[Dict[str, Any]] = Field(
+        title="Properties to aid the invocation of the processor on the annotated code. If none are provided, then the Snowflake CLI will try to get the current environment information.",
```
Comment (sfc-gh-bdufour) [inline src/snowflake/cli/api/project/schemas/native_app/path_mapping.py:14]: Ditto, this should be generic. Something like "Set of key-value pairs used to configure the output of the processor. Consult a specific processor's documentation for more details on supported properties."

**Review by sfc-gh-bdufour — DISMISSED** [review]

**Pull request merged**